### ISSUE TITLE
*proof* CoD Mobile works, needs keyboard input.

### ISSUE THREAD

**Issue by papermache** [Genymobile/scrcpy#841]
A little laggy but great graphics and works wirelessly. Mouse and touchscreen also work, would like keyboard instead to control the game. ![WorkSpaces](https://user-images.githubusercontent.com/2126012/66261630-23ec0f00-e79f-11e9-9e21-0d9fb87bc0f5.png)

**Comment by rom1v** [comment]
The keyboard does not work in the game?

**Comment by papermache** [comment]
If it's thru an emulator keyboard and mouse are assigned. Not sure where to assign bindings with scrcpy.

**Comment by volcano1111** [comment]
> The keyboard does not work in the game?
Yeah, it doesn't work. I tried to press all of the buttons, arrows, etc - no luck. If you could make this thing working it would be very convenient in the games.

**Comment by papermache** [comment]
I use Barrier https://github.com/debauchee/barrier on Linux so somewhere in that codebase must be keyboard support for scrcpy.

**Comment by papermache** [comment]
Keyboard and mouse support should come from a Barrier APK, not scrcpy.

**Comment by rom1v** [comment]
Do the keys work with `adb shell input keyevent ...`?

**Comment by rom1v** [comment]
What does `adb logcat` says when you press keys?

**Comment by volcano1111** [comment]
Thanks for reopening this thread :)

> 
> 
> What does `adb logcat` says when you press keys?

Here's the output when I tried to press some keys in the main menu: [](https://pastebin.com/U1gsMykF). The only key that worked in my case was ESC (keyCode:111). It also works when I'm typing **.\adb.exe shell input keyevent 111**. The other keys are not working in any case, neither by pressing, nor by typing **.\adb.exe shell input keyevent**.

**Comment by rom1v** [comment]
Then I would say that the game just did not implement keyboard controls.

**Comment by volcano1111** [comment]
> 
> 
> Then I would say that the game just did not implement keyboard controls.

Well, I can say that gamepad is not working either. Probably it was done intentionally, so that people can't play on a PC.

**Comment by EthraZa** [comment]
Something like 
`scrcpy shell input keyevent` (that uses the already running server of scrcpy instead of adb laggy way)
would be way cool and useful.

**Comment by papermache** [comment]
Nvidia Geforce Now apk works on android with built-in multi controller and keyboard so emulating laptop keyboard support thru scrcpy would enable game control from the laptop instead of Android.

**Comment by Flight-X** [comment]
would you help me to do that for cod mobile
